pkg/controller/validatingadmissionpolicystatus/crd.go

```diff
@@ -0,0 +1,256 @@
+limitations under the License.
+*/
+
+package validatingadmissionpolicystatus
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+	"time"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionsv1client "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/cel/openapi/resolver"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+type crdTracker struct {
+	// affectedPolicies is the (CRD GroupResources) -> (affected policy names) mapping
+	affectedPolicies map[schema.GroupResource]sets.Set[string]
+	// policyDependencies is the (policy name) -> (required CRD GroupResource) mapping.
+	// It is the reverse of affectedPolicies.
+	policyDependencies map[string]sets.Set[schema.GroupResource]
+
+	// mu protects affectedPolicies and policyDependencies
+	mu sync.RWMutex
+
+	parent *Controller
+}
+
+// convertCRDValidationToSchema converts the openAPIV3Schema to a kube-openapi schema.
+// In case of nil validation or missing OpenAPIV3Schema, this function returns ErrSchemaNotFound.
+// The CustomResourceValidation object must be valid. Anything from the informer or client should have already passed
+// validation.
+// This function uses a JSON marshal-unmarshal conversion instead of Structural because
+//   - OpenAPIV3Schema is already validated and thus always a valid OpenAPI schema;
+//   - Structural involves private api-extensions API types. We have import-boss to prevent anything outside
+//     apiextensions-apiserver package to use them.
+func convertCRDValidationToSchema(validation *apiextensionsv1.CustomResourceValidation) (*spec.Schema, error) {
+	if validation == nil || validation.OpenAPIV3Schema == nil {
+		return nil, resolver.ErrSchemaNotFound
+	}
+	b, err := json.Marshal(validation.OpenAPIV3Schema)
+	if err != nil {
+		return nil, err
+	}
+	s := new(spec.Schema)
+	err = json.Unmarshal(b, s)
+	if err != nil {
+		return nil, err
+	}
+	return s, nil
+}
+
+func newCRDTracker(controller *Controller) *crdTracker {
+	return &crdTracker{
+		affectedPolicies:   make(map[schema.GroupResource]sets.Set[string]),
+		policyDependencies: make(map[string]sets.Set[schema.GroupResource]),
+		parent:             controller,
+	}
+}
+
+// handleCRDChange handles incoming changes to CRDs.
+// CRDs are not enqueued. Instead, policy definitions that are affected by the changed CRDs are enqueued to policyQueue.
+func (t *crdTracker) handleCRDChange(crd *apiextensionsv1.CustomResourceDefinition) {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	for _, gr := range []schema.GroupResource{
+		{
+			Group:    crd.Spec.Group,
+			Resource: crd.Spec.Names.Plural,
+		},
+		{
+			Group:    crd.Spec.Group,
+			Resource: crd.Spec.Names.Singular,
```

Comment:
Note: the name of CRD limits to Spec.Names.Plural + Spec.Group  https://github.com/kubernetes/kubernetes/blob/c575a1e31aa60126d200b7137a130b91ecb8bb36/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L77
Not sure if singular is needed. I am ok with having both just to be safe

Reply:
Updated to make it consistent.